src/memory/managed_ptr.hpp

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <cuda.h>
+
+#include <memory/allocator.hpp>
+
+namespace nest {
+namespace mc {
+namespace memory {
+
+// used to indicate that the type pointed to by the managed_ptr is to be
+// constructed in the managed_ptr constructor
+struct construct_in_place_tag {};
+
+// Like std::unique_ptr, but for CUDA managed memory.
+// Handles memory allocation and freeing, and the construction and destruction
+// of the type being stored in the allocated memory.
+// Implemented as a stand alone type instead of as a std::unique_ptr with
+// custom desctructor so that __device__ annotation can be added to members
+// like ::get, ::operator*, etc., which enables the use of the smart pointer
+// in device side code.
+//
+// It is very strongly recommended that the helper make_managed_ptr be used
+// instead of directly constructing the managed_ptr.
+template <typename T>
+class managed_ptr {
+    public:
+
+    using element_type = T;
+    using pointer = element_type*;
+    using reference = element_type&;
+
+    managed_ptr() = default;
+
+    managed_ptr(const managed_ptr& other) = delete;
+
+    // Allocate memory and construct in place using args.
+    // This is an extension over the std::unique_ptr interface, because the
+    // point of the wrapper is to hide the complexity of allocating managed
+    // memory and constructing a type in place.
+    template <typename... Args>
+    managed_ptr(construct_in_place_tag, Args&&... args) {
+        managed_allocator<element_type> allocator;
+        data_ = allocator.allocate(1u);
```

Comment:
Hmmmm, maybe I have deviated from the C++ standard allocator interface. The allocator is templated on type `T`, so asking for `1u` is the same as `1u*sizeof(T)`.
So, no, it is golden. But, might want to standardise allocator interfaces inside the memory namespace.